File: read_bfee.c
void read_bfee(unsigned char *inBytes, mxArray *outCell)
{
	unsigned int timestamp_low = inBytes[0] + (inBytes[1] << 8) +
		(inBytes[2] << 16) + (inBytes[3] << 24);
	unsigned short bfee_count = inBytes[4] + (inBytes[5] << 8);
	unsigned int Nrx = inBytes[8];
	unsigned int Ntx = inBytes[9];
	unsigned int rssi_a = inBytes[10];
	unsigned int rssi_b = inBytes[11];
	unsigned int rssi_c = inBytes[12];
	char noise = inBytes[13];
	unsigned int agc = inBytes[14];
	unsigned int antenna_sel = inBytes[15];
	unsigned int len = inBytes[16] + (inBytes[17] << 8);
	unsigned int fake_rate_n_flags = inBytes[18] + (inBytes[19] << 8);
	unsigned int calc_len = (30 * (Nrx * Ntx * 8 * 2 + 3) + 7) / 8;
	unsigned int i, j;
	unsigned int index = 0, remainder;
	unsigned char *payload = &inBytes[20];
	char tmp;
	mwSize size[] = {Ntx, Nrx, 30};
	mxArray *csi = mxCreateNumericArray(3, size, mxDOUBLE_CLASS, mxCOMPLEX);
	mwSize perm_size[] = {1, 3};
	mxArray *perm = mxCreateNumericArray(2, perm_size, mxDOUBLE_CLASS, mxREAL);
	double* ptrR = (double *)mxGetPr(csi);
	double* ptrI = (double *)mxGetPi(csi);

	/* Check that length matches what it should */
	if (len != calc_len)
		mexErrMsgIdAndTxt("MIMOToolbox:read_bfee_new:size","Wrong beamforming matrix size.");

	/* Compute CSI from all this crap :) */
	for (i = 0; i < 30; ++i)
	{
		index += 3;
		remainder = index % 8;
		for (j = 0; j < Nrx * Ntx; ++j)
		{
			tmp = (payload[index / 8] >> remainder) |
				(payload[index/8+1] << (8-remainder));
			//printf("%d\n", tmp);
			*ptrR = (double) tmp;
			++ptrR;
			tmp = (payload[index / 8+1] >> remainder) |
				(payload[index/8+2] << (8-remainder));
			*ptrI = (double) tmp;
			++ptrI;
			index += 16;
		}
	}

	/* Compute the permutation array */
	ptrR = (double *)mxGetPr(perm);
	ptrR[0] = ((antenna_sel) & 0x3) + 1;
	ptrR[1] = ((antenna_sel >> 2) & 0x3) + 1;
	ptrR[2] = ((antenna_sel >> 4) & 0x3) + 1;
}

File: read_bfee.py
import numpy as np

def usig2sig(b):
    if b > 127:
        return b
    else:
        return ~(b-128) + 1

def read_bfee(Bytes):
    timestamp_low = int.from_bytes(Bytes[0:4], 'little')
    bfee_count = int.from_bytes(Bytes[4:6], 'little')
    Nrx = Bytes[8]
    Ntx = Bytes[9]
    rssi_a = Bytes[10]
    rssi_b = Bytes[11]
    rssi_c = Bytes[12]
    noise = usig2sig(Bytes[13])
    agc = Bytes[14]
    antenna_sel = Bytes[15]
    len = int.from_bytes(Bytes[16:18], 'little')
    fake_rate_n_flags = int.from_bytes(Bytes[18:20], 'little')
    calc_len = int((30 * (Nrx * Ntx * 8 * 2 + 3) + 7) / 8)
    index = 0
    payload = Bytes[20:]
    csi = np.empty((Ntx, Nrx, 30), dtype=complex, order = 'C')
    perm = [((antenna_sel) & 0x3) + 1, ((antenna_sel >> 2) & 0x3) + 1, ((
            antenna_sel >> 4) & 0x3) + 1]
    if len != calc_len:
        print('Error: Wrong beamforming matrix size.')
        return
    
    for i in range(0, 30):
        index += 3
        remainder = index % 8;
        for j in range(0, Ntx):
            for k in range(0, Nrx):
                re = (payload[(int)(index/8)] >> remainder) | (payload[(int)(index/8)+1] << (8-remainder))
			       #printf("%d\n", tmp);
                im = (payload[(int)(index/8)+1] >> remainder) | (payload[(int)(index/8)+2] << (8-remainder))
                csi[j,k,i] = complex(re, im)
                index += 16
                   
 
    structure =[timestamp_low, bfee_count, Nrx, Ntx, rssi_a, rssi_b, rssi_c, 
                noise, agc, perm, fake_rate_n_flags, csi]
    return structure

'''                  
    st = np.dtype([('timestamp_low', 'I'), ('bfee_count', 'H'),
                   ('Nrx', 'B'), ('Ntx', 'B'), 
                   ('rssi_a', 'B'), ('rssi_b', 'B'), ('rssi_c', 'B'),
                   ('noise', 'B'), ('agc', 'B'),
                   ('perm', 'B'), ('rate', 'H'),
                   ('csi', 'complex64')])
    structure = np.array([(timestamp_low, bfee_count,
                           Nrx, Ntx, rssi_a, rssi_b, rssi_c,
                           noise, agc, perm, fake_rate_n_flags, csi)], dtype = st)
'''